crates/kubelet/src/pod/status.rs

```diff
@@ -153,46 +142,83 @@ pub fn make_status_with_containers(
     reason: &str,
     container_statuses: Vec<KubeContainerStatus>,
     init_container_statuses: Vec<KubeContainerStatus>,
-) -> serde_json::Value {
-    serde_json::json!(
-       {
-           "metadata": {
-               "resourceVersion": "",
-           },
-           "status": {
-               "phase": phase,
-               "reason": reason,
-               "containerStatuses": container_statuses,
-               "initContainerStatuses": init_container_statuses,
-           }
-       }
-    )
+) -> Status {
+    let mut status = Status::new();
+    status.set_phase(phase);
+    status.set_reason(reason);
+    status.set_container_statuses(container_statuses);
+    status.set_init_container_statuses(init_container_statuses);
+    status
 }
 
-/// Describe the status of a workload.
 #[derive(Clone, Debug, Default)]
-pub struct Status {
-    /// Allows a provider to set a custom message, otherwise, kubelet will infer
-    /// a message from the container statuses
-    pub message: StatusMessage,
-    /// The statuses of containers keyed off their names
-    pub container_statuses: ContainerMap<ContainerStatus>,
-}
+/// Pod Status wrapper.
+pub struct Status(KubePodStatus);
 
-#[derive(Clone, Debug)]
-/// The message to be set in a pod status update.
-pub enum StatusMessage {
-    /// Do not change the existing status message.
-    LeaveUnchanged,
-    /// Remove any existing status message.
-    Clear,
-    /// Set the status message to the given value.
-    Message(String),
-}
+impl Status {
+    /// Create a new status with no fields set.
+    pub fn new() -> Self {
+        Status(Default::default())
+    }
 
-impl Default for StatusMessage {
-    fn default() -> Self {
-        Self::LeaveUnchanged
+    /// Set Pod phase.
+    pub fn set_phase(&mut self, phase: Phase) {
+        self.0.phase = Some(format!("{}", phase));
+    }
+
+    /// Set Pod reason.
+    pub fn set_reason(&mut self, reason: &str) {
+        self.0.reason = Some(reason.to_string());
+    }
+
+    /// Set Pod message.
+    pub fn set_message(&mut self, message: &str) {
+        self.0.message = Some(message.to_string());
+    }
+
+    /// Set Pod container statuses.
+    pub fn set_container_statuses(&mut self, container_statuses: Vec<KubeContainerStatus>) {
+        self.0.container_statuses = Some(container_statuses);
+    }
+
+    /// Set Pod init container statuses.
+    pub fn set_init_container_statuses(
+        &mut self,
+        init_container_statuses: Vec<KubeContainerStatus>,
+    ) {
+        self.0.init_container_statuses = Some(init_container_statuses);
+    }
+
+    pub(crate) fn into_json(self) -> serde_json::Value {
+        let mut status = serde_json::Map::new();
```

Comment:
See above. Calling `json!` doesnt really create a patch. It may be that what we really want is to use `jsondiff` here, but then you need to have an updated version of the Pod status and then there are race conditions, and this actually seems simpler.

Reply:
Ah, ok valid point